### ISSUE TITLE
StorageClass header is not added to upload request

### ISSUE THREAD

**Issue by padamuscinova** [aws/aws-sdk-android#375]
When using TransferUtility for upload, request (I'm using single part only) is sent without StorageClass header although it's set in objectmetadata:

```
ObjectMetadata metadata = new ObjectMetadata();
                metadata.setContentType(MIME_TYPE);
                metadata.setHeader(Headers.STORAGE_CLASS, StorageClass.StandardInfrequentAccess);

transferObserver = transferUtility.upload(AWSConfig.getBucketId(),
                String.format(KEY_FORMAT, item.getUuid()),
                new File(item.getUrl()),
                metadata,
                CannedAccessControlList.PublicRead);
```
As far as I've tried to read the source code, it seems it's not even read from the object metadata/stored in aws sdk local transfer db and thus it can't be added as header when doing the actual request.

Tested with SDK version: 2.6.6, 2.6.7

**Comment by kvasukib** [comment]
Hi Petra,

Thank you for reporting to us. The setHeader method is meant for the internal purpose and not for public consumption. The TransferUtility doesn't honor the headers set in setHeader API of `ObjectMetadata`.

Instead, you can pass in the custom metadata using the following API:
https://github.com/aws/aws-sdk-android/blob/master/aws-android-sdk-s3/src/main/java/com/amazonaws/services/s3/model/ObjectMetadata.java#L177

You need to create a map and enter your key-value pairs and pass on the map to `setCustomMetadata` API in`ObjectMetadata`.

**Comment by padamuscinova** [comment]
Hello,

thank you for your answer. I've tried the suggested method but unfortunately it adds  `x-amz-meta-x-amz-storage-class: STANDARD_IA` header instead of just `x-amz-storage-class: STANDARD_IA` (iOS sends the header without the x-meta-meta-)

This is my code now:
```
ObjectMetadata metadata = new ObjectMetadata();
                metadata.setContentType(MIME_TYPE);
                metadata.setUserMetadata(Collections.singletonMap(Headers.STORAGE_CLASS, StorageClass.StandardInfrequentAccess.toString()));

transferObserver = transferUtility.upload(AWSConfig.getBucketId(),
                String.format(KEY_FORMAT, item.getUuid()),
                new File(item.getUrl()),
                metadata,
                CannedAccessControlList.PublicRead);
```

**Comment by gabrielkaputa** [comment]
Hi,

we are having the same issue and it's really annoying because we have to set the storage class on the backend, for that we need to copy the object and that is causing duplicate sns notifications of new objects in s3. Is there any news when this will be fixed?

Thanks

**Comment by kvasukib** [comment]
Hi @gabrielkaputa, Have you tried the source code mentioned by @padamuscinova? Does that work for you?

**Comment by gabrielkaputa** [comment]
Hi, I did, same problem. The storage class is not set.

**Comment by padamuscinova** [comment]
@kvasukib It can’t work as it doesn’t work for me either - the header is set but in invalid format.

**Comment by padamuscinova** [comment]
#398

**Comment by gabrielkaputa** [comment]
how about some activity instead? setting storage class seems like a pretty important thing

**Comment by kvasukib** [comment]
I was able to reproduce the issue. The issue is that the S3 putObject method extracts the storage class from the put object request and adds it as an Http header to the underlying request object. However, the TransferUtility does not have a provision to accept the storage class directly and we rely on the customer user metadata which is added to the `ObjectMetadata`. The UploadTask in TransferUtility does not add the storage class to the put object request.

I made a change to add the storage class to the put object request when it's been part of the custom user-metadata and found it to be working for the 4 storage classes except for Glacier. 

@padamuscinova Were you able to set Glacier storage class and test it?
